src/xdp/program.c

```diff
@@ -11,10 +11,163 @@
+        NonPagedPoolNx, sizeof(EBPF_PROG_TEST_RUN_CONTEXT), XDP_POOLTAG_PROGRAM);
+    if (XdpMd->ProgTestRunContext == NULL) {
+        EbpfResult = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+
+    // Allocate buffer for data.
+    XdpMd->ProgTestRunContext->Data = (char*)ExAllocatePoolZero(NonPagedPoolNx, DataSizeIn, XDP_POOLTAG_PROGRAM);
+    if (XdpMd->ProgTestRunContext->Data == NULL) {
+        EbpfResult = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+    memcpy(XdpMd->ProgTestRunContext->Data, DataIn, DataSizeIn);
+    XdpMd->ProgTestRunContext->DataSize = DataSizeIn;
+
+    XdpMd->Base.data = (void*)XdpMd->ProgTestRunContext->Data;
+    XdpMd->Base.data_end = (void*)(XdpMd->ProgTestRunContext->Data + XdpMd->ProgTestRunContext->DataSize);
+
+    if (context_in != NULL && ContextSizeIn >= sizeof(xdp_md_t)) {
+        xdp_md_t* xdp_context = (xdp_md_t*)context_in;
+        XdpMd->Base.data_meta = xdp_context->data_meta;
+        XdpMd->Base.ingress_ifindex = xdp_context->ingress_ifindex;
+    }
+
+    *Context = XdpMd;
+    XdpMd = NULL;
+
+    EbpfResult = EBPF_SUCCESS;
+
+Exit:
+    if (XdpMd != NULL) {
+        if (XdpMd->ProgTestRunContext != NULL) {
+            if (XdpMd->ProgTestRunContext->Data != NULL) {
+                ExFreePool(XdpMd->ProgTestRunContext->Data);
+            }
+            ExFreePool(XdpMd->ProgTestRunContext);
+        }
+        ExFreePool(XdpMd);
+    }
+
+    return EbpfResult;
+}
+
+static void
+XdpDeleteContext(
+    _In_opt_ void* Context,
+    _Out_writes_bytes_to_(*DataSizeOut, *DataSizeOut) uint8_t* DataOut,
+    _Inout_ size_t* DataSizeOut,
+    _Out_writes_bytes_to_(*ContextSizeOut, *ContextSizeOut) uint8_t* ContextOut,
+    _Inout_ size_t* ContextSizeOut)
+{
+    EBPF_XDP_MD* XdpMd = NULL;
+
+    if (!Context) {
+        goto Exit;
+    }
+
+    XdpMd = (EBPF_XDP_MD*)Context;
+
+    // Copy the packet data to the output buffer.
+    if (DataOut != NULL && DataSizeOut != NULL && XdpMd->Base.data != NULL) {
+        size_t DataSize = *DataSizeOut;
+        size_t XdpDataSize = (char*)(XdpMd->Base.data_end) - (char*)(XdpMd->Base.data);
+        if (DataSize > XdpDataSize) {
+            DataSize = XdpDataSize;
+        }
+        memcpy(DataOut, XdpMd->Base.data, DataSize);
+        *DataSizeOut = DataSize;
+    } else {
+        *DataSizeOut = 0;
+    }
+
+    // Copy some fields from the Context to the output buffer.
+    if (ContextOut != NULL && ContextSizeOut != NULL) {
+        size_t context_size = *ContextSizeOut;
+        if (context_size > sizeof(xdp_md_t)) {
+            context_size = sizeof(xdp_md_t);
+        }
+
+        xdp_md_t* XdpContextOut = (xdp_md_t*)ContextOut;
+        XdpContextOut->data_meta = XdpMd->Base.data_meta;
+        XdpContextOut->ingress_ifindex = XdpMd->Base.ingress_ifindex;
+        *ContextSizeOut = context_size;
+    } else {
+        *ContextSizeOut = 0;
+    }
+
+    if (XdpMd->ProgTestRunContext != NULL) {
```

Comment:
Ideally this code and the error handling cleanup code in the context create function could be in one common subroutine.